administrator/components/com_installer/tmpl/downloadkeys/default.php

```diff
@@ -0,0 +1,124 @@
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+JHtml::_('behavior.multiselect');
+
+JHtml::_('bootstrap.tooltip');
+
+$listOrder = $this->escape($this->state->get('list.ordering'));
+$listDirn  = $this->escape($this->state->get('list.direction'));
+?>
+<div id="installer-manage" class="clearfix">
+	<form action="<?php echo JRoute::_('index.php?option=com_installer&view=downloadkeys'); ?>" method="post" name="adminForm" id="adminForm">
+		<div class="row">
+			<div id="j-sidebar-container" class="col-md-2">
+				<?php echo $this->sidebar; ?>
+			</div>
+			<div class="col-md-10">
+				<div id="j-main-container" class="j-main-container">
+					<?php echo JLayoutHelper::render('joomla.searchtools.default', array('view' => $this)); ?>
+					<?php if (empty($this->items)) : ?>
+					<div class="alert alert-warning alert-no-items">
+						<?php echo JText::_('JGLOBAL_NO_MATCHING_RESULTS'); ?>
+					</div>
+					<?php else : ?>
+					<table class="table table-striped">
+						<thead>
+							<tr>
+								<th style="width:1%" class="text-center">
+									<?php echo JHtml::_('grid.checkall'); ?>
+								</th>
+								<th class="nowrap">
+									<?php echo JHtml::_('searchtools.sort', 'COM_INSTALLER_HEADING_UPDATESITE_NAME', 'update_site_name', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:20%" class="nowrap hidden-sm-down text-center">
+									<?php echo JHtml::_('searchtools.sort', 'COM_INSTALLER_HEADING_NAME', 'name', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:10%" class="hidden-sm-down text-center">
+									<?php echo JHtml::_('searchtools.sort', 'COM_INSTALLER_HEADING_LOCATION', 'client_translated', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:10%" class="hidden-sm-down text-center">
+									<?php echo JHtml::_('searchtools.sort', 'COM_INSTALLER_HEADING_TYPE', 'type_translated', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:10%" class="hidden-sm-down text-center">
+									<?php echo JHtml::_('searchtools.sort', 'COM_INSTALLER_HEADING_FOLDER', 'folder_translated', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:5%" class="nowrap hidden-sm-down text-center">
+									<?php echo JHtml::_('searchtools.sort', 'JGRID_HEADING_ID', 'update_site_id', $listDirn, $listOrder); ?>
+								</th>
+								<th style="width:5%" class="nowrap hidden-sm-down text-center">
+									<?php echo JText::_('COM_INSTALLER_HEADING_DOWNLOADKEY'); ?>
+								</th>
+							</tr>
+						</thead>
+						<tfoot>
+							<tr>
+								<td colspan="8">
+									<?php echo $this->pagination->getListFooter(); ?>
+								</td>
+							</tr>
+						</tfoot>
+						<tbody>
+						<?php $itemsSkipped = 0; ?>
+						<?php foreach ($this->items as $i => $item) : ?>
+							<?php $extraQuery = InstallerHelper::getExtraQuery($item->update_site_id);?>
```

Comment:
There will need to be a more efficient way to handle this.  Right now you are triggering one extra database query for every item in the collection, plus parsing the extension XML from the filesystem.